# Predicting NHL Team's Chances at Qualifying for the Playoffs

Maggie Dong (14618573), David Wilson (88190509)

#### Introduction

By American Thanksgiving, approximately a quarter into the 82 game season, statistics show that 84% of the NHL teams currently sitting in a playoff spot will retain their playoff berth at the season’s end. The goal of this project is to look at the current season’s (2022-23) standings and predict the probability that each team will make the playoffs based on historical data.

In [10]:
# Libraries
import pandas as pd

#### Formulate a Dataset

In [17]:
# Important! This notebook needs to be in a folder along with a folder named "Data" containing the datasets.

# Import Data
# x_2017 = pd.read_excel(r'Data\2017_20_games.xlsx')
# x_2018 = pd.read_excel(r'Data\2018_20_games.xlsx')
# x_2019 = pd.read_excel(r'Data\2019_20_games.xlsx')
# x_2020 = pd.read_excel(r'Data\2020_(Jan-Feb_2021)_20_games.xlsx')
# x_2021 = pd.read_excel(r'Data\2021_20_games.xlsx')

# y_2017 = pd.read_excel(r'Data\2017_18_full_season.xlsx')
# y_2018 = pd.read_excel(r'Data\2018_19_full_season.xlsx')
# y_2019 = pd.read_excel(r'Data\2019_20_full_season.xlsx')
# y_2020 = pd.read_excel(r'Data\2020_21_full_season.xlsx')
# y_2021 = pd.read_excel(r'Data\2021_22_full_season.xlsx')

x_2017 = pd.read_excel(r'Data/2017_20_games.xlsx')
x_2018 = pd.read_excel(r'Data/2018_20_games.xlsx')
x_2019 = pd.read_excel(r'Data/2019_20_games.xlsx')
x_2020 = pd.read_excel(r'Data/2020_(Jan-Feb_2021)_20_games.xlsx')
x_2021 = pd.read_excel(r'Data/2021_20_games.xlsx')

y_2017 = pd.read_excel(r'Data/2017_18_full_season.xlsx')
y_2018 = pd.read_excel(r'Data/2018_19_full_season.xlsx')
y_2019 = pd.read_excel(r'Data/2019_20_full_season.xlsx')
y_2020 = pd.read_excel(r'Data/2020_21_full_season.xlsx')
y_2021 = pd.read_excel(r'Data/2021_22_full_season.xlsx')

# Include a year column for each x df for reference
x_2017.insert(1, 'Year', 2017)
x_2018.insert(1, 'Year', 2018)
x_2019.insert(1, 'Year', 2019)
x_2020.insert(1, 'Year', 2020)
x_2021.insert(1, 'Year', 2021)

# Isolate P% column of y df's to form response vector
y_2017 = y_2017['P%']
y_2018 = y_2018['P%']
y_2019 = y_2019['P%']
y_2020 = y_2020['P%']
y_2021 = y_2021['P%']

# Stack x's together and y's respectively to form 1 dataset
x = pd.concat([x_2017, x_2018, x_2019, x_2020, x_2021])
y = pd.concat([y_2017, y_2018, y_2019, y_2020, y_2021])

In [18]:
x.head()

Team  Year  GP   W  L   T  OT   P     P%  RW  ...  GA  \
0    Tampa Bay Lightning  2017  21  16  3  --   2  34  0.810  13  ...  53   
1        St. Louis Blues  2017  22  16  5  --   1  33  0.750  12  ...  58   
2    Toronto Maple Leafs  2017  23  14  8  --   1  29  0.630   9  ...  68   
3  Columbus Blue Jackets  2017  22  14  7  --   1  29  0.659   7  ...  52   
4          Winnipeg Jets  2017  21  13  5  --   3  29  0.690  12  ...  57   

   GF/GP  GA/GP   PP%   PK%  Net PP%  Net PK%  Shots/GP  SA/GP  FOW%  
0   3.90   2.52  27.2  83.3     25.9     88.9      32.8   31.9  47.7  
1   3.45   2.64  16.7  78.6     13.9     80.0      33.6   30.5  48.7  
2   3.48   2.96  21.9  80.8     19.2     80.8      31.1   33.8  51.3  
3   2.77   2.36  10.0  83.6      6.7     83.6      34.9   30.8  46.6  
4   3.24   2.71  22.9  78.5     21.4     83.5      28.3   33.8  51.3  

[5 rows x 23 columns]

In [4]:
len(x), len(y)

(156, 156)

We expect x and y to be of length 156 since there have been 4 seasons with 31 teams and 1 with 32. $4*31+32=156$.

So, it looks like everything has transferred over nicely and we didnt lose any of the data along the way!

#### Feature Selection